### PR TITLE
Implements pytato.loopy.call_loopy

### PR DESCRIPTION
**Pull request by kaushikcfd**
- Branched from #16 
- Adds support for calling loopy kernels

TODO:
- undo change to https://github.com/inducer/pytato/pull/20/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03f

### CODE REVIEW

**Comment by kaushikcfd** [comment]
Pylint failures are tricky to handle cleany, as most of the `pytato.call_loopy` depends on interfaces that aren't part of loopy's master branch yet and adding those many "ignore" annotations would be quite noisy. Happy to hear if there's a better option.

File: test/test_codegen.py [hunk 1]
```diff
@@ -497,6 +506,108 @@ def test_math_functions(ctx_factory, dtype, function_name):
...
+    x = pt.make_placeholder(ns, shape="(M, N)", dtype=np.float, name="x")
+    y = 3*x
+
+    knl = lp.make_kernel(
+            "{[i, j]: 0<=i<m and 0<=j<n}",
+            """
+            Z[i] = 7*sum(j, Y[i, j])
+            """, name="callee", lang_version=(2018, 2))
+
+    loopyfunc = call_loopy(ns, knl, bindings={"Y": y, "m": "M", "n": "N"})
```
Comment (inducer) [inline test/test_codegen.py:603]: To `**kwargs` or not to `**kwargs`? Maybe have a kwargs'd version?

Comment (inducer) [inline test/test_codegen.py:603]: Shape inference? (OK to bump to issue for later)

File: pytato/array.py [hunk 2]
```diff
@@ -755,25 +784,32 @@ class DictOfNamedArrays(Mapping[str, Array]):
...
+    def __getitem__(self, name: str) -> NamedArray:
+        return self._named_arrays[name]
 
     def __iter__(self) -> Iterator[str]:
-        return iter(self._data)
+        return iter(self._named_arrays)
 
     def __len__(self) -> int:
-        return len(self._data)
+        return len(self._named_arrays)
```
Comment (inducer) [inline pytato/array.py:808]: Why this change of traversal behavior? This broke `pytato.visualization` (in the sense that that now only shows the `NamedArray`s, which isn't useful).

cc @VincentWells

Comment (inducer) [reply pytato/array.py:808]: @VincentWells It seems like this diff gets the visualizer going again:
```diff
diff --git a/pytato/visualization.py b/pytato/visualization.py
index c4f22f1..09aff97 100644
--- a/pytato/visualization.py
+++ b/pytato/visualization.py
@@ -201,7 +201,7 @@ def get_dot_graph(result: Union[Array, DictOfNamedArrays]) -> str:
 
     nodes: Dict[Array, DotNodeInfo] = {}
     mapper = ArrayToDotNodeInfoMapper()
-    for elem in itertools.chain(namespace.values(), outputs.values()):
+    for elem in itertools.chain(namespace.values(), outputs._data.values()):
         mapper(elem, nodes)
 
     input_arrays: List[Array] = []
@@ -239,7 +239,7 @@ def get_dot_graph(result: Union[Array, DictOfNamedArrays]) -> str:
                 emit('%s -> %s [label="%s"]' % (tail, head, dot_escape(label)))
 
         # Emit output/namespace name mappings.
-        _emit_name_cluster(emit, outputs, array_to_id, id_gen, label="Outputs")
+        _emit_name_cluster(emit, outputs._data, array_to_id, id_gen, label="Outputs")
         _emit_name_cluster(emit, namespace, array_to_id, id_gen, label="Namespace")
 
     return emit.get()
```
There are still a big pile of (mostly unused) `DataWrapper`s and their names that clutter up the graph, but maybe you can figure out a way to filter them out. (PR welcome!)

@kaushikcfd Any clue what's up with those unused `DataWrapper`s that hang out in the namespace?

Comment (inducer) [inline pytato/array.py:808]: @kaushikcfd Any clue what's up with those unused `DataWrapper`s that hang out in the namespace?

Comment (kaushikcfd) [reply pytato/array.py:808]: Are you pointing at the unused DataWrappers in the DG operator DAG?
If so, it maybe that there are many calls of this pattern: https://github.com/inducer/meshmode/blob/9bf8f9f1b77aeb2f7eaff18aaa5f4e66dbaf2986/meshmode/discretization/connection/direct.py#L207.

I.e. DataWrappers get frozen and again thawed.

Comment (kaushikcfd) [reply pytato/array.py:808]: > Why this change of traversal behavior? This broke pytato.visualization (in the sense that that now only shows the NamedArrays, which isn't useful).

We had to separate out results of a DictOfNamedArrays from the expression inside it. Slightly odd but helps in the a more concrete definition of a LoopyFunction (sub-classed from DictOfNamedArrays) and LoopyFunctionResult (sub-classed from NamedArray).

**Comment by inducer** [comment]
This picked up a  few conflicts.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

**Comment by matthiasdiener** [comment]
I think this is ready for another review @inducer @kaushikcfd .

File: pytato/array.py [hunk 3]
```diff
@@ -626,7 +626,32 @@ class CountNamed(UniqueTag):
 
 # {{{ dict of named arrays
 
-class DictOfNamedArrays(Mapping[str, Array]):
+class NamedArray(Array):
```
Comment (inducer) [inline pytato/array.py:629]: Add docstring, include in docs?

File: pytato/loopy.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+.. autoclass:: LoopyFunctionResult
+
+.. autofunction:: call_loopy
+"""
+
+
+class LoopyFunction(DictOfNamedArrays):
+    """
+    Call to a :mod:`loopy` program.
```
Comment (inducer) [inline pytato/loopy.py:52]:
```suggestion
    An array expression node representing a call to an entrypoint in a
    :mod:`loopy` translation unit.
```

File: pytato/loopy.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+.. autoclass:: LoopyFunction
+
+.. autoclass:: LoopyFunctionResult
+
+.. autofunction:: call_loopy
+"""
+
+
+class LoopyFunction(DictOfNamedArrays):
```
Comment (inducer) [inline pytato/loopy.py:50]: Add to docs.

File: pytato/array.py [hunk 6]
```diff
@@ -642,21 +667,28 @@ class DictOfNamedArrays(Mapping[str, Array]):
         This container deliberately does not implement
         arithmetic.
     """
```
Comment (inducer) [inline pytato/array.py:669]: Updated docs a bit for `NamedArray`.

File: pytato/loopy.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        loopy_arg = self.dict_of_named_arrays.entry_kernel.arg_dict[  # type:ignore
+                self.name]
+        dtype = loopy_arg.dtype
+
+        if isinstance(dtype, np.dtype):
+            return dtype
+        elif isinstance(dtype, NumpyType):
+            return np.dtype(dtype.numpy_dtype)
+        else:
+            raise NotImplementedError(f"Unknown dtype type '{dtype}'")
```
Comment (inducer) [inline pytato/loopy.py:130]:
```suggestion
        return loopy_arg.dtype.numpy_dtype
```

File: pytato/loopy.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        if ((self.entrypoint == other.entrypoint)
+             and (self.bindings == other.bindings)
+             and (self.program == other.program)):
+            return True
+        return False
+
+
+class LoopyFunctionResult(NamedArray):
+    """
+    Named array for :class:`LoopyFunction`'s result.
```
Comment (inducer) [inline pytato/loopy.py:100]:
```suggestion
    Named array for :class:`LoopyFunction`'s result.
    Inherits from :class:`~pytato.array.NamedArray`.
```

File: pytato/loopy.py (hunk 5 above)
Comment (inducer) [inline pytato/loopy.py:50]:
```suggestion
class LoopyCall(DictOfNamedArrays):
```

File: pytato/array.py [hunk 9]
```diff
@@ -642,21 +667,28 @@ class DictOfNamedArrays(Mapping[str, Array]):
...
+    _mapper_method = "map_dict_of_named_arrays"
 
     def __init__(self, data: Dict[str, Array]):
+        self._named_arrays = {name: NamedArray(self, name)
+                              for name in data}
         self._data = data
```
Comment (inducer) [inline pytato/array.py:675]: `LoopyFunction` (`LoopyCall`) should call this superclass constructor. (Maybe just allow supplying `_named_arrays`?)

Comment (inducer) [inline pytato/array.py:675]:
```suggestion
        # kept around mainly for efficient __hash__
        self._data = data
```
(if that's true... otherwise `_data` is pretty redundant with `_named_arrays`.)

File: pytato/loopy.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+        self.program = program
+        self.bindings = bindings
+        self.entrypoint = entrypoint
+
+        entry_kernel = program[entrypoint]
+
+        self._named_arrays = {name: LoopyFunctionResult(self, name)
+                              for name, lp_arg in entry_kernel.arg_dict.items()
+                              if lp_arg.is_output}
```
Comment (inducer) [inline pytato/loopy.py:70]: Should set `_named_arrays` via the superclass constructor.

File: pytato/loopy.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+    @memoize_method
+    def to_pytato(self, expr: ScalarExpression) -> ScalarExpression:
```
Comment (inducer) [inline pytato/loopy.py:73]: Underscore? (I'm guessing this is not meant to be user-facing.)

File: pytato/loopy.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        from pymbolic.mapper.substitutor import make_subst_func
+        return SubstitutionMapper(make_subst_func(self.bindings))(expr)
+
+    @property
+    def entry_kernel(self) -> lp.LoopKernel:
```
Comment (inducer) [inline pytato/loopy.py:78]: Underscore? (I'm guessing this is not meant to be user-facing.)

File: pytato/loopy.py [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+def call_loopy(program: "lp.Program",
+               bindings: Dict[str, Union[Array, Number]],
+               entrypoint: Optional[str] = None) -> LoopyFunction:
+    """
+    Operates a general :class:`loopy.Program` on the array inputs as specified
+    by *bindings*.
+
+    Restrictions on the structure of ``program[entrypoint]``:
+
+    * array arguments of ``program[entrypoint]`` should either be either
```
Comment (inducer) [inline pytato/loopy.py:142]:
```suggestion
    * array arguments of ``program[entrypoint]`` must either be either
```

File: pytato/loopy.py [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+        if isinstance(dtype, np.dtype):
+            return dtype
+        elif isinstance(dtype, NumpyType):
+            return np.dtype(dtype.numpy_dtype)
+        else:
+            raise NotImplementedError(f"Unknown dtype type '{dtype}'")
+
+
+def call_loopy(program: "lp.Program",
```
Comment (inducer) [inline pytato/loopy.py:133]: Add to docs.

Comment (inducer) [inline pytato/loopy.py:133]: Program -> TranslationUnit

File: pytato/loopy.py [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+
+
+class LoopyFunction(DictOfNamedArrays):
+    """
+    Call to a :mod:`loopy` program.
+    """
+    _mapper_method = "map_loopy_function"
+
+    def __init__(self,
+            program: "lp.Program",
```
Comment (inducer) [inline pytato/loopy.py:57]: Program -> TranslationUnit (Attribute should also be renamed)

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -34,6 +34,7 @@ jobs:
                 python-version: '3.x'
         -   name: "Main Script"
             run: |
+                sed -i "s/loopy.git/loopy.git@kernel_callables_v3-edit2/g" requirements.txt
```
Comment (inducer) [inline .github/workflows/ci.yml:37]: File an issue so we remember to point this back to main.

**Review by inducer — COMMENTED** [review]
Thanks! This looks great. Just a few (mostly cosmetic) things.

File: pytato/array.py (hunk 3 above)
Comment (matthiasdiener) [reply pytato/array.py:630]: Added. Correct like it is @kaushikcfd ?

File: pytato/array.py (hunk 6 above)
Comment (matthiasdiener) [reply pytato/array.py:674]: Updated.

File: pytato/loopy.py (hunk 15 above)
Comment (matthiasdiener) [reply pytato/loopy.py]: Renamed.

File: pytato/loopy.py (hunk 12 above)
Comment (matthiasdiener) [reply pytato/loopy.py]: added underscore

File: pytato/loopy.py (hunk 11 above)
Comment (matthiasdiener) [reply pytato/loopy.py]: added underscore

File: pytato/loopy.py (hunk 14 above)
Comment (matthiasdiener) [reply pytato/loopy.py]: changed

Comment (matthiasdiener) [reply pytato/loopy.py]: added

File: pytato/loopy.py (hunk 5 above)
Comment (matthiasdiener) [reply pytato/loopy.py]: added

File: pytato/loopy.py (hunk 10 above)
Comment (matthiasdiener) [reply pytato/loopy.py]: Done (I think).

File: pytato/array.py (hunk 9 above)
Comment (matthiasdiener) [reply pytato/array.py]: Done (I think).

**Comment by matthiasdiener** [comment]
> Thanks! This looks great. Just a few (mostly cosmetic) things.

Thanks! I think I addressed your comments @inducer.

**Comment by inducer** [comment]
> Apply suggestions from code review

:( Please use commit messages that'll still make sense from the perspective of the commit history later on. Those commits won't get squashed away.

File: pytato/loopy.py [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+    _mapper_method = "map_loopy_function"
+
+    def __init__(self,
+            translation_unit: "lp.TranslationUnit",
+            bindings: Dict[str, Union[Array, Number]],
+            entrypoint: str):
+        entry_kernel = translation_unit[entrypoint]
+        super().__init__({name: LoopyCallResult(self, name)
+                              for name, lp_arg in entry_kernel.arg_dict.items()
+                              if lp_arg.is_output})
```
Comment (inducer) [inline pytato/loopy.py:63]: I think these will get "double-wrapped" now, once in `LoopyCallResult` and again in `NamedArray`. I think if you want to pass `NamedArray`s to the superclass, you'll need to make a separate argument.

Comment (inducer) [inline pytato/loopy.py:63]: Assuming I'm right about the double-wrapping, how come none of the tests noticed this?

**Review by inducer — COMMENTED** [review]
Thanks! A few more comments below.

File: pytato/array.py [hunk 18]
```diff
@@ -642,21 +672,29 @@ class DictOfNamedArrays(Mapping[str, Array]):
         This container deliberately does not implement
         arithmetic.
     """
+    _mapper_method = "map_dict_of_named_arrays"
 
     def __init__(self, data: Dict[str, Array]):
+        self._named_arrays = {name: NamedArray(self, name)
+                              for name in data}
+        # kept around mainly for efficient __hash__
         self._data = data
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/array.py:682]: "kept around for efficient `__hash__`", I'm guessing that was for "_data" attribute, then not quite. A NamedArray comprises of a DictOfNamedArray and it's name in it i.e. it's more of a reference in the DictOfNamedArray. However, `_data` is the one which actually contains the array expressions. `NamedArray.expr` might explain the need for this.

The reason that a NamedArray doesn't contain the array expression is because NamedArray shouldn't be anything by itself, as it is always comes in a group. Take for example a `LoopyFunctionResult`, we cannot emit expressions that would only compute one of the outputs of a `LoopyFunction` i.e. each of them is just a reference to in a group of results.

Probably I should add docs highlighting this if you agree with the decision taken here of keeping things centralized at DictOfNamedArrays.

File: pytato/loopy.py (hunk 17 above)
Comment (kaushikcfd) [reply pytato/loopy.py:63]: Oh, yep these shouldn't be "double-wrapped'. Didn't get caught in the tests as the mappers for the named array just point back all the operations to it's DictOfNamedArrays. Will fix this.

Comment (kaushikcfd) [reply pytato/loopy.py:63]: Continuing the discussion at https://github.com/inducer/pytato/pull/20/files#r645290705

File: pytato/loopy.py [hunk 19]
```diff
@@ -112,66 +112,66 @@ def expr(self) -> Array:
...
                 self.name]
-        shape: ShapeType = self.dict_of_named_arrays.to_pytato(  # type:ignore
+        shape: ShapeType = self.dict_of_named_arrays._to_pytato(  # type:ignore
                 loopy_arg.shape)
         return shape
 
     @property
-    def dtype(self) -> np.dtype[Any]:
-        loopy_arg = self.dict_of_named_arrays.entry_kernel.arg_dict[  # type:ignore
+    def dtype(self) -> Any:
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/loopy.py:119]: We always emit `Any` for dtype.

File: pytato/loopy.py [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+import loopy as lp
+import pymbolic.primitives as prim
+from typing import Dict, Optional, Any, Union
+from numbers import Number
+from pytato.array import DictOfNamedArrays, Array, ShapeType, NamedArray
+from pytato.scalar_expr import SubstitutionMapper, ScalarExpression
+from pytools import memoize_method
+from pytools.tag import TagsType
+
+__doc__ = """
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/loopy.py:38]: I think it is in the docs (reference.rst)

File: pytato/loopy.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/loopy.py:119]: A decision that was chosen quite early on, and I had inferred that's because to accomodate for "whatever 'np.dtype' can take in.

[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/loopy.py:119]: (side note - mypy complained about np.dtype[Any])

File: pytato/array.py (hunk 3 above)
Comment (kaushikcfd) [reply pytato/array.py:630]: The docs are accurate.

File: pytato/loopy.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/loopy.py]: I think we had forgot to wrap it in numpy dtype: c7e5875

File: pytato/array.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (inducer) [reply pytato/array.py:682]: Thanks for explaining! I was clearly confused here; there's no redundancy at all. I think a little bit of explanation might be useful here, along the lines of "There are two sets of nodes here. _data contains the expressions used to populate the container, whereas _named_arrays contains array nodes that are used by expressions depending on values in the container."

File: pytato/loopy.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply pytato/loopy.py:38]: Added a nitpick argument to sphinx which will warn about these

File: doc/Makefile [hunk 21]
```diff
@@ -3,7 +3,7 @@
 
 # You can set these variables from the command line, and also
 # from the environment for the first two.
-SPHINXOPTS    ?=
+SPHINXOPTS    ?= -n
```
Comment (inducer) [inline doc/Makefile:6]: There shouldn't be a need. `build_docs` adds this by default:

https://github.com/inducer/ci-support/blob/4edcdff1f7126d437630cdd0a6e2551a4ec5fb59/ci-support.sh#L423-L427

Comment (matthiasdiener) [reply doc/Makefile:6]: Ok, but isn't it convenient for local builds to gave this enabled?

Comment (inducer) [reply doc/Makefile:6]: Then
```suggestion
SPHINXOPTS    ?= -W --keep-going -n
```
?

Comment (matthiasdiener) [reply doc/Makefile:6]: Well, it currently does not build with `-W -n`, at least on my system, and independent of this PR.

Comment (inducer) [reply doc/Makefile:6]: 78b9aec

Comment (inducer) [reply doc/Makefile:6]: (Long story short here: I had disabled doc checking because pre-4.0 sphinx couldn't handle the type annotations here. That commit reenables them.)

File: pytato/array.py [hunk 22]
```diff
@@ -626,38 +619,103 @@ class CountNamed(UniqueTag):
 
 # {{{ dict of named arrays
 
-class DictOfNamedArrays(Mapping[str, Array]):
-    """A container that maps valid Python identifiers
-    to instances of :class:`Array`. May occur as a result
-    type of array computations.
+class NamedArray(Array):
+    """An entry in a :class:`DictOfNamedArrays`. Holds a reference back to the
+    containing instance as well as the name by which *self* is known there.
```
Comment (inducer) [inline pytato/array.py:624]:
```suggestion
    """An entry in a :class:`AbstractResultWithNamedArrays`. Holds a reference
    back to thecontaining instance as well as the name by which *self* is
    known there.
```

File: pytato/array.py [hunk 23]
```diff
@@ -626,38 +619,103 @@ class CountNamed(UniqueTag):
...
+    @property
+    def dtype(self) -> np.dtype[Any]:
+        return self.expr.dtype
+
+
+class AbstractResultWithNamedArrays(Mapping[str, NamedArray], ABC):
+    r"""An abstract array computation that results in multiple :class:`Array`\ s,
+    each named. The way in which the values of these arrays are computed
+    is determined by concrete subclasses of this class, e.g.
+    :class:`pytato.loopy.LoopyCall`.
```
Comment (inducer) [inline pytato/array.py:659]:
```suggestion
    :class:`pytato.loopy.LoopyCall` or :class:`DictOfNamedArrays`.
```

**Comment by inducer** [comment]
LGTM. Thanks to both of you for working on this!
